### PR TITLE
Add initial checkin for openebs chart

### PR DESCRIPTION
**Pull request by kmova**
Signed-off-by: kmova <kiran.mova@openebs.io>

*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

### CODE REVIEW

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR. Here's an initial quick review.

* Please make sure you follow our guidelines (e. g. names and labels)
* Please apply our current RBAC best practices: https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md
* Compare with what `helm create` using Helm 2.8+ gives you and update accordingly.

**Comment by kmova** [comment]
Thank you @unguiculus I have updated the PR with some changes to reflect the label and value (conventions)

File: incubator/openebs/templates/cm-openebs-prometheus-config.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+{{- if .Values.policies.monitoring.enabled }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "openebs.fullname" . }}-prometheus-config
+{{- if .Values.rbac.create }}
+  namespace: {{ .Values.rbac.namespace }}
```
Comment (unguiculus) [inline incubator/openebs/templates/cm-openebs-prometheus-config.yaml]: Why do you install resources into a different namespace depending on whether or not RBAC is enabled?

File: incubator/openebs/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+## To uninstall/delete the `my-release` deployment:
+```
+helm ls --all
+helm delete `my-release`
+```
+
+
+## Configuration
+
```
Comment (unguiculus) [inline incubator/openebs/README.md]: The table needs an update.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/openebs/README.md (hunk 2 above)
Comment (kmova) [reply incubator/openebs/README.md]: Thanks @unguiculus - updated the table.

File: incubator/openebs/templates/cm-openebs-prometheus-config.yaml (hunk 1 above)
Comment (kmova) [reply incubator/openebs/templates/cm-openebs-prometheus-config.yaml]: Good point. This was used to enable/disable service account. I removed this check and kept it only for creating and associating service account.

File: incubator/openebs/templates/deployment-maya-apiserver.yaml [hunk 3]
```diff
@@ -0,0 +1,79 @@
+apiVersion: apps/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "openebs.fullname" . }}-maya-apiserver
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-maya-apiserver.yaml]: Suffix and component label should be the same.

File: incubator/openebs/templates/deployment-maya-apiserver.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+spec:
+  replicas: {{ .Values.apiserver.replicas }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "openebs.name" . }}
+        release: {{ .Release.Name }}
+        component: apiserver
+    spec:
+{{- if .Values.rbac.create }}
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-maya-apiserver.yaml]: Remove if block and use `{{ template "openebs.serviceAccountName". }}`

File: incubator/openebs/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+{{- end -}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+Create chart name and version as used by the chart label.
+*/}}
+{{- define "openebs.chart" -}}
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
```
Comment (unguiculus) [inline incubator/openebs/templates/_helpers.tpl]: Please add the named template for the service account as suggested at https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md.

File: incubator/openebs/templates/deployment-maya-apiserver.yaml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+{{ toYaml .Values.apiserver.nodeSelector | indent 8 }}
+{{- end }}
+
+{{- if .Values.apiserver.tolerations }}
+      tolerations:
+{{ toYaml .Values.apiserver.tolerations | indent 8 }}
+{{- end }}
+
+{{- if .Values.apiserver.antiAffinity.enabled }}
+      affinity:
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-maya-apiserver.yaml]: For more flexibility, do this:

```yaml
{{- if .Values.apiserver.affinity }}
      affinity:
{{ toYaml .Values.apiserver.affinity | indent 8 }}
{{- end }}
```

This would also be consistent with tolerations and nodeSelector.

File: incubator/openebs/templates/deployment-maya-provisioner.yaml [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+spec:
+  replicas: {{ .Values.provisioner.replicas }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "openebs.name" . }}
+        release: {{ .Release.Name }}
+        component: provisioner
+    spec:
+{{- if .Values.rbac.create }}
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-maya-provisioner.yaml]: Remove if block and use `{{ template "openebs.serviceAccountName". }}`

File: incubator/openebs/templates/deployment-openebs-prometheus.yaml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: {{ template "openebs.name" . }}
+        release: {{ .Release.Name }}
+        component: prometheus
+    spec:
+{{- if .Values.rbac.create }}
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-openebs-prometheus.yaml]: Remove if block and use `{{ template "openebs.serviceAccountName". }}`

File: incubator/openebs/templates/cm-openebs-prometheus-config.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
+{{- if .Values.policies.monitoring.enabled }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "openebs.fullname" . }}-prometheus-config
+  namespace: {{ .Values.rbac.namespace }}
```
Comment (unguiculus) [inline incubator/openebs/templates/cm-openebs-prometheus-config.yaml]: You should probably remove the namespace from the resources. Also, I don't see how this is related to RBAC. The namespace can be specified by the user when they install the chart.

File: incubator/openebs/templates/namespace.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
```
Comment (unguiculus) [inline incubator/openebs/templates/namespace.yaml]: Remove namespace resource

File: incubator/openebs/templates/deployment-openebs-prometheus.yaml [hunk 11]
```diff
@@ -0,0 +1,75 @@
+{{- if .Values.policies.monitoring.enabled }}
+apiVersion: apps/v1beta1
+kind: Deployment
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-openebs-prometheus.yaml]: There's a Prometheus chart. You should use a dependency to that instead.

File: incubator/openebs/templates/deployment-openebs-grafana.yaml [hunk 12]
```diff
@@ -0,0 +1,39 @@
+{{- if .Values.policies.monitoring.enabled }}
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-openebs-grafana.yaml]: There's a Grafana chart. You should use a dependency to that instead.

File: incubator/openebs/templates/sc-openebs-cassandra.yaml [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# Define a storage classes supported by OpenEBS
```
Comment (unguiculus) [inline incubator/openebs/templates/sc-openebs-cassandra.yaml]: Why do you create so many storage classes, especially for things that are completely unrelated to this chart? A storage class is something that is usually not managed within the scope of a chart. This has usually a broader scope. A chart would then just reference storage classes. Also, I don't understand why you create application-specific storage classes.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/openebs/templates/deployment-openebs-prometheus.yaml (hunk 11 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-openebs-prometheus.yaml]: Thanks @unguiculus. It makes sense. I have removed the prometheus and grafana related files.  In fact, there is no dependency on the prometheus/grafana for openebs to run. If the user wants to monitor openebs via prometheus/grafana, they could add the configuration jobs/dashboards into their already existing instances or use their respective charts to install and configure .

File: incubator/openebs/templates/deployment-openebs-grafana.yaml (hunk 12 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-openebs-grafana.yaml]: removed this.

File: incubator/openebs/templates/cm-openebs-prometheus-config.yaml (hunk 9 above)
Comment (kmova) [reply incubator/openebs/templates/cm-openebs-prometheus-config.yaml]: this file is removed.

File: incubator/openebs/templates/deployment-openebs-prometheus.yaml (hunk 8 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-openebs-prometheus.yaml]: this file is removed

File: incubator/openebs/templates/sc-openebs-cassandra.yaml (hunk 13 above)
Comment (kmova) [reply incubator/openebs/templates/sc-openebs-cassandra.yaml]: OpenEBS is like an storage infrastructure service. Depending on the type of stateful application, the OpenEBS can provide/provision PVs that are best suited for an application. This can be accomplished by creating application specific storage classes with customized properties. While this is the intent of having multiple storage classes, your observation is right! As of today most of these classes look the same and there is no need to create so many classes. 

I will only keep - openebs-standard (that has replication=3) and openebs-standalone (replication=1). In the upcoming releases, as new features are added specific to application - for example auto-snapshoting for database, new classes can be created.

Or we could just document how the storage classes can be created and let the users create their own storage class.

File: incubator/openebs/templates/deployment-maya-apiserver.yaml (hunk 3 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-maya-apiserver.yaml]: done! Thank you!

File: incubator/openebs/templates/_helpers.tpl (hunk 5 above)
Comment (kmova) [reply incubator/openebs/templates/_helpers.tpl]: done

File: incubator/openebs/templates/deployment-maya-provisioner.yaml (hunk 7 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-maya-provisioner.yaml]: Removing the if block is failing my test. When I run the test with rbac.create=false, the template still fills in the serviceAccountName which is not created.

File: incubator/openebs/templates/namespace.yaml (hunk 10 above)
Comment (kmova) [reply incubator/openebs/templates/namespace.yaml]: done! thank you!

**Comment by kmova** [comment]
@unguiculus - still working on few more comments, which I will try to push with my next commit:
- refactor the "affinity" section
- refactor to remove the if blocks around serviceAccount, still trying to find out if this could be done. 
- update the readme

File: incubator/openebs/templates/deployment-maya-apiserver.yaml (hunk 6 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-maya-apiserver.yaml]: done

File: incubator/openebs/templates/deployment-maya-provisioner.yaml (hunk 7 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-maya-provisioner.yaml]: My bad, the test was failing due to the bug - I was checking for rbac.create for creating serviceAccount. Fixed it.

**Comment by kmova** [comment]
Hi @unguiculus. incorporated the comments. thank you!

**Comment by kmova** [comment]
ping @unguiculus  .. Looking forward to your feedback.

**Comment by muratkars** [comment]
Hi @unguiculus. Checking if you have any changes required after @kmova 's comments. Thank you!

File: incubator/openebs/Chart.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+description: Containerized Storage for Containers
+icon: https://raw.githubusercontent.com/openebs/chitrakala/master/OpenEBS%20logo/openebs%20logos-03.png
+home: http://www.openebs.io/
+keywords:
+  - cloud-native-storage
+  - block-storage
+  - iSCSI
+  - storage
+sources:
+  - https://github.com/openebs/openebs
```
Comment (unguiculus) [inline incubator/openebs/Chart.yaml]: Add maintainers

File: incubator/openebs/templates/clusterrole.yaml [hunk 15]
```diff
@@ -0,0 +1,29 @@
+{{- if .Values.rbac.create }}
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: {{ template "openebs.serviceAccountName". }}
```
Comment (unguiculus) [inline incubator/openebs/templates/clusterrole.yaml]: {{ template "openebs.fullname" . }}

File: incubator/openebs/templates/clusterrolebinding.yaml [hunk 16]
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.rbac.create }}
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  name: {{ template "openebs.serviceAccountName". }}
```
Comment (unguiculus) [inline incubator/openebs/templates/clusterrolebinding.yaml]: {{ template "openebs.fullname" . }}

File: incubator/openebs/templates/clusterrolebinding.yaml [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+  labels:
+    app: {{ template "openebs.name" . }}
+    chart: {{ template "openebs.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: {{ template "openebs.serviceAccountName". }}
```
Comment (unguiculus) [inline incubator/openebs/templates/clusterrolebinding.yaml]: {{ template "openebs.fullname" . }}

File: incubator/openebs/templates/deployment-maya-apiserver.yaml [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        #- name: OPENEBS_IO_K8S_MASTER
+        #  value: "http://172.28.128.3:8080"
+        - name: OPENEBS_IO_JIVA_CONTROLLER_IMAGE
+          value: "{{ .Values.jiva.image }}:{{ .Values.jiva.imageTag }}"
+        - name: OPENEBS_IO_JIVA_REPLICA_IMAGE
+          value: "{{ .Values.jiva.image }}:{{ .Values.jiva.imageTag }}"
+        - name: OPENEBS_IO_VOLUME_MONITOR_IMAGE
+          value: "{{ .Values.policies.monitoring.image }}:{{ .Values.policies.monitoring.imageTag }}"
+        - name: OPENEBS_IO_JIVA_REPLICA_COUNT
+          value: "{{ .Values.jiva.replicas }}"
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-maya-apiserver.yaml]: Remove empty lines

File: incubator/openebs/templates/deployment-maya-provisioner.yaml [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+          value: "{{ template "openebs.fullname" . }}-apiservice"
+        # The following values will be set as annotations to the PV object.
+        # Refer : https://github.com/openebs/external-storage/pull/15
+        #- name: OPENEBS_MONITOR_URL
+        #  value: "{{ .Values.provisioner.monitorUrl }}"
+        #- name: OPENEBS_MONITOR_VOLKEY
+        #  value: "{{ .Values.provisioner.monitorVolumeKey }}"
+        #- name: MAYA_PORTAL_URL
+        #  value: "{{ .Values.provisioner.mayaPortalUrl }}"
+
```
Comment (unguiculus) [inline incubator/openebs/templates/deployment-maya-provisioner.yaml]: Remove empty lines

File: incubator/openebs/templates/sc-openebs-cassandra.yaml (hunk 13 above)
Comment (unguiculus) [reply incubator/openebs/templates/sc-openebs-cassandra.yaml]: I think the storage classes you create are very specific and may not suit a user's needs. The chart itself doesn't need them. Users should create their own storage classes as they need them. It'd be better to just document things.

File: incubator/openebs/templates/clusterrole.yaml (hunk 15 above)
Comment (kmova) [reply incubator/openebs/templates/clusterrole.yaml]: done

File: incubator/openebs/templates/clusterrolebinding.yaml (hunk 16 above)
Comment (kmova) [reply incubator/openebs/templates/clusterrolebinding.yaml]: thanks. done

File: incubator/openebs/templates/clusterrolebinding.yaml (hunk 17 above)
Comment (kmova) [reply incubator/openebs/templates/clusterrolebinding.yaml]: done. thanks!

File: incubator/openebs/templates/deployment-maya-apiserver.yaml (hunk 18 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-maya-apiserver.yaml]: removed. thanks!

File: incubator/openebs/templates/deployment-maya-provisioner.yaml (hunk 19 above)
Comment (kmova) [reply incubator/openebs/templates/deployment-maya-provisioner.yaml]: removed. thanks!

**Comment by rtluckie** [comment]
@kmova @unguiculus any movement on this PR?

File: incubator/openebs/values.yaml [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+rbac:
+  # Specifies whether RBAC resources should be created
+  create: true
+
+serviceAccount:
+  create: true
+  name: "openebs-maya-operator"
```
Comment (unguiculus) [inline incubator/openebs/values.yaml]: Default should be empty

**Review by unguiculus — CHANGES_REQUESTED** [review]
I had suggested remove the storage classes and document how to create them instead. would you follow this advice?

Then I'd suggest you move the chart to stable directly.

File: incubator/openebs/values.yaml (hunk 20 above)
Comment (kmova) [reply incubator/openebs/values.yaml]: done

**Comment by kmova** [comment]
Removed the storage classes and updated NOTES.txt with details on creating storage class.

**Comment by kmova** [comment]
@unguiculus - moved openebs to stable after incorporating the changes. Thanks!

**Review by unguiculus — APPROVED** [review]
